presto-hive/src/main/java/com/facebook/presto/hive/metastore/MetastoreUtil.java

```diff
@@ -0,0 +1,354 @@
+            if (!first) {
+                colNameBuf.append(",");
+                colTypeBuf.append(":");
+                colComment.append('\0');
+            }
+            colNameBuf.append(col.getName());
+            colTypeBuf.append(col.getType());
+            colComment.append(col.getComment().orElse(""));
+            first = false;
+        }
+        String colNames = colNameBuf.toString();
+        String colTypes = colTypeBuf.toString();
+        schema.setProperty(hive_metastoreConstants.META_TABLE_COLUMNS, colNames);
+        schema.setProperty(hive_metastoreConstants.META_TABLE_COLUMN_TYPES, colTypes);
+        schema.setProperty("columns.comments", colComment.toString());
+
+        schema.setProperty(serdeConstants.SERIALIZATION_DDL, toThriftDDL(tableName, dataColumns));
+
+        String partString = "";
+        String partStringSep = "";
+        String partTypesString = "";
+        String partTypesStringSep = "";
+        for (Column partKey : partitionKeys) {
+            partString = partString.concat(partStringSep);
+            partString = partString.concat(partKey.getName());
+            partTypesString = partTypesString.concat(partTypesStringSep);
+            partTypesString = partTypesString.concat(partKey.getType().getHiveTypeName());
+            if (partStringSep.length() == 0) {
+                partStringSep = "/";
+                partTypesStringSep = ":";
+            }
+        }
+        if (partString.length() > 0) {
+            schema.setProperty(hive_metastoreConstants.META_TABLE_PARTITION_COLUMNS, partString);
+            schema.setProperty(hive_metastoreConstants.META_TABLE_PARTITION_COLUMN_TYPES, partTypesString);
+        }
+
+        if (parameters != null) {
+            for (Map.Entry<String, String> e : parameters.entrySet()) {
+                // add non-null parameters to the schema
+                if (e.getValue() != null) {
+                    schema.setProperty(e.getKey(), e.getValue());
+                }
+            }
+        }
+
+        return schema;
+    }
+
+    private static String toThriftDDL(String structName, List<Column> columns)
+    {
+        // Mimics function in Hive:
+        // MetaStoreUtils.getDDLFromFieldSchema(String, List<FieldSchema>)
+        StringBuilder ddl = new StringBuilder();
+        ddl.append("struct ");
+        ddl.append(structName);
+        ddl.append(" { ");
+        boolean first = true;
+        for (Column col : columns) {
+            if (first) {
+                first = false;
+            }
+            else {
+                ddl.append(", ");
+            }
+            ddl.append(typeToThriftType(col.getType().getHiveTypeName()));
+            ddl.append(' ');
+            ddl.append(col.getName());
+        }
+        ddl.append("}");
+        return ddl.toString();
+    }
+
+    private static ProtectMode getProtectMode(Map<String, String> parameters)
+    {
+        if (parameters == null) {
```

Comment:
I don't think this happens since the source of this call is our metadata object and they never allow null.